@here/harp-geometry/lib/SubdivisionModifier.ts

```diff
@@ -49,83 +62,74 @@ export abstract class SubdivisionModifier {
             }
 
             // the position of the new vertex.
-            const p = new Vector3();
-            p.lerpVectors(vertices[i], vertices[j], 0.5);
+            tmpVectorA.set(position[i * 3], position[i * 3 + 1], position[i * 3 + 2]);
+            tmpVectorB.set(position[j * 3], position[j * 3 + 1], position[j * 3 + 2]);
+            tmpVectorC.lerpVectors(tmpVectorA, tmpVectorB, 0.5);
 
             // the index of the new vertex.
-            const index = vertices.length;
-            vertices.push(p);
+            const index = position.length / 3;
+            position.push(...tmpVectorC.toArray());
             // cache the position of the new vertex.
             cache.set(key, index);
 
-            return index;
-        }
-
-        // The resulting triangles.
-        const newFaces: Face3[] = [];
-        const newFaceVertexUvs: Vector2[][] = [];
+            if (uv !== undefined) {
+                tmpVectorA.set(uv[i * 2], uv[i * 2 + 1], 0);
+                tmpVectorB.set(uv[j * 2], uv[j * 2 + 1], 0);
+                tmpVectorC.lerpVectors(tmpVectorA, tmpVectorB, 0.5);
+                uv.push(tmpVectorC.x, tmpVectorC.y);
+            }
 
-        const uvWorkList = oldUvs[0];
-        const hasUvs = oldUvs !== undefined && oldUvs.length > 0 && uvWorkList.length > 0;
+            if (edge !== undefined) {
+                if (edge[i] === j) {
+                    edge.push(j);
+                    edge[i] = index;
+                } else if (edge[j] === i) {
+                    edge.push(i);
+                    edge[j] = index;
+                } else {
+                    edge.push(-1);
+                }
+            }
 
-        while (true) {
-            const face = faceWorkList.shift();
+            return index;
+        }
 
-            if (face === undefined) {
+        const newIndices = [];
+        while (indices.length >= 3) {
+            const v0 = indices.shift();
+            const v1 = indices.shift();
+            const v2 = indices.shift();
+            if (v0 === undefined || v1 === undefined || v2 === undefined) {
```

Comment:
Why should this ever happen if the length is >= 3 ?

Reply:
Right, leftover from previous implementation. Good catch.